### ISSUE TITLE
Training Zipformer using fp16 leads to divergence??

### ISSUE THREAD

**Issue by ZQuang2202** [k2-fsa/icefall#1462]
### Discussed in https://github.com/k2-fsa/icefall/discussions/1461

<div type='discussions-op-text'>

<sup>Originally posted by **ZQuang2202** January 14, 2024</sup>
Hi everyone,
I am a student attempting to reproduce the results of Zipformer on Librispeech 100h, facing limitations in hardware resources that prevent me from using the recommended configuration. Due to these constraints, I have reduced the batch size (max_duration) to 300, as opposed to the recommended 1000. However, I am struggling to find the appropriate configuration for Eden.

1. Following the training strategy that suggests decreasing the learning rate by √k times when the batch size decreases by k times, I initially set the base_lr to 0.03, keeping other configurations at their default values. But the training process diverges. Despite attempts to adjust lr_batches, lr_epochs (3.5-6), and base_lr (0.3-0.45), it's still not working. Notably, the training process encounters divergence when the batch_count is around 700-900, leading to 'parameter domination' issues in the embed_conv and some attention modules. I attach some log information below.
![image](https://github.com/k2-fsa/icefall/assets/152836329/5e15854b-ab8f-4b6c-9226-7604dba0db1f)
![image](https://github.com/k2-fsa/icefall/assets/152836329/5d36af68-134d-4c43-a235-ba3d3354db3f)
![image](https://github.com/k2-fsa/icefall/assets/152836329/a5c580e9-8969-48f4-a679-2e687b51c04f)

2. In an effort to address this, I attempted to reduce the gradient scale of the layers experiencing 'parameter domination,' but this proved ineffective."

I have few questions: 
1. Is there some thing related between base_lr/lr_batch/lr_epoch and max_duration? 
2. Should parameters like dropout rate or min_abs, max_abs of Balancer or Whitener's config be changed? If so, how should it be changed for each different dataset?
Thank you.</div>

**Comment by sangeet2020** [comment]
I have somewhat similar issue: https://github.com/k2-fsa/icefall/issues/1463#issuecomment-1893492095

**Comment by marcoyang1998** [comment]
One way to alleviate this issue is to use gradient accumulation to simulate large batch sizes. This will require a bit of extra code in `train.py`.

**Comment by danpovey** [comment]
Mm, I think it would be more straightforward to just decrease the --initial-lr.

**Comment by yaozengwei** [comment]
In my test, the training goes well with a base-lr of 0.025, for max-duration=300 and 1 GPU.
